Tribal Section 1 Parsing & Validation

elipe17 opened the pull request:
## Summary of Changes
- Added support for Tribal TANF section 1 parsing and validation
- Added new pre-check error that manages tribe code header inconsistencies
Pull request closes #1114 

## How to Test
```
cd tdrs-frontend && docker-compose up
cd tdrs-backend && docker-compose up
```

- Open http://localhost:3000/ and sign in.
- Submit both `ADS.E2J.FTP1.TS142` and `tribal_section_1_inconsistency.txt`.
- Verify the first file generates Tribal T1/T2/T3 records.
- Verify the second file does not generate any records and throws a pre-check error describing an inconsistency with the tribe code and other header fields.

## Deliverables
_More details on how deliverables herein are assessed included [here](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
- [x] Data File parsing logic is in place such that data file submission can be completed by user and parsed by system for full submission once errors are resolved.
- [ ] Data Errors tell user specifically where in the file they occur, and therefore what exactly needs to be fixed in order for the submission to be complete
- [x] TDP gives the user all data errors for a submission in a way they can easily save or export so they can take the issues back to their case management software and fix the issues then recreate an export for resubmission to TDP.
- [ ] Data Errors must be issued in human readable plain language so that an English speaker can easily understand what is needed. 
- [x] Testing Checklist has been run and all tests pass
- [x] README is updated, if necessary

### [Deliverable 2: Tested Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
  + [x] If this PR introduces backend code changes, are they meaningfully tested?
  + [ ] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [x] Frontend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)
  + [x] Backend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)

### [Deliverable 3: Properly Styled Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)

+ [x] Are backend code style checks passing on CircleCI?
+ [x] Are frontend code style checks passing on CircleCI?
+ [x] Are code maintainability principles being followed?

### [Deliverable 4: Accessible](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)

+ [ ] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [ ] Does PR include documentation for Raft's a11y review? 
+ [ ] Did automated and manual testing with `iamjolly` and `ttran-hub` using Accessibility Insights reveal any errors introduced in this PR?


### [Deliverable 5: Deployed](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)

+ [x] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### [Deliverable 6: Documented](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)

+ [ ] Does this PR provide background for why coding decisions were made?
+ [x] If this PR introduces backend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces frontend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?
+ [x] Can reviewer explain and take ownership of these elements presented in this code review?

### [Deliverable 7: Secure](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)

+ [x] Does the OWASP Scan pass on CircleCI?
+ [ ] Do manual code review and manual testing detect any new security issues?
+ [ ] If new issues detected, is investigation and/or remediation plan documented? 

### [Deliverable 8: User Research](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
+ [ ] the purpose of the research
+ [ ] methods used to conduct the research 
+ [ ] who participated in the research
+ [ ] what was tested and how
+ [ ] impact of research on TDP
+ [ ] (_if applicable_) final design mockups produced for TDP development

raftmsohani committed:
Added formating for header and autofit columns

raftmsohani committed:
Formatted the headers

raftmsohani committed:
added year/month to the columns

raftmsohani committed:
Added contants - translation column

raftmsohani committed:
added friendly names to T1 and T2

raftmsohani committed:
added friendly name to m1 and m2

raftmsohani committed:
added friendly name to m3

raftmsohani committed:
added friendly_name to t3

raftmsohani committed:
added friendly_name to t4 and t5

raftmsohani committed:
added friendly_name to t7

raftmsohani committed:
correct missing friendly_name

raftmsohani committed:
correction on failing tests

raftmsohani committed:
addedfriendly name to excel report

raftmsohani committed:
linting

raftmsohani committed:
linting

raftmsohani committed:
linting

raftmsohani committed:
delete contants.py

raftmsohani committed:
added test for json field in error model

raftmsohani committed:
linting

raftmsohani committed:
linting

raftmsohani committed:
linting

elipe17 committed:
- Added Tribal models, docs, admin refs
- Added support for tribal parsing

elipe17 committed:
- Update t1 field validators

elipe17 committed:
- t2/t3 field validators

raftmsohani committed:
2599-added friendly name to postparsing validators

elipe17 committed:
- Updated cat3 validators

elipe17 committed:
- Updated how we detect tribe code

elipe17 committed:
- Added error handling for inconsistent header info wrt tribe code
- Added new test

elipe17 committed:
- Updated test

elipe17 committed:
- Grammar fix
- Removed unused function

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into 1114-tribal-s1

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into 1114-tribal-s1

raftmsohani committed:
refining the validator tests

elipe17 committed:
- Format with Black
- Add missing tests

elipe17 committed:
- Fix lint

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into 1114-tribal-s1

raftmsohani committed:
added returning fields names to validators

raftmsohani committed:
added friendly_name to error field

raftmsohani committed:
Merge branch 'develop' into 2599-error-report-corrections-impact-wpr-metrics

raftmsohani committed:
linting

raftmsohani committed:
corrections on views/tests

raftmsohani committed:
corrections for fields

raftmsohani committed:
failing test corrected

raftmsohani committed:
failing test corrected

raftmsohani committed:
correcting test failures

raftmsohani committed:
linting

raftmsohani committed:
corrected the excel fiel generator

andrew-jameson commented on tdrs-backend/tdpservice/parsers/parse.py:
```diff
@@ -28,11 +28,29 @@ def parse_datafile(datafile):
+                                                                        util.make_generate_parser_error(datafile, 1))
+
+    if not tribe_is_valid:
+        logger.info("Tribe Code binconsistency with Program Type and FIPS Code.")
```
```suggestion
        logger.info("Tribe Code inconsistency with Program Type and FIPS Code.")
```

andrew-jameson commented on tdrs-backend/tdpservice/parsers/parse.py:
```diff
@@ -117,7 +135,7 @@ def rollback_parser_errors(datafile):
     logger.debug(f"Deleted {num_deleted} {ParserError}.")
 
-def parse_datafile_lines(datafile, program_type, section, is_encrypted):
+def parse_datafile_lines(datafile, program_type, section, is_encrypted, is_tribal):
```
Why do we pass in `is_tribal` if it appears to be unused in this function? Also, isn't it a little redundant if we can determine that later from the datafile obj?

andrew-jameson commented on tdrs-backend/tdpservice/parsers/schema_defs/ssp/m1.py:
```diff
@@ -8,101 +8,402 @@
+                validators.hasLength(150),
+            ],
+            postparsing_validators=[],
+            fields=[
```
What was the driving change for the reformatting of all these SSP files? Bloats the PR a bit.

andrew-jameson commented on tdrs-backend/tdpservice/parsers/schema_defs/tanf/t2.py:
```diff
@@ -9,235 +9,878 @@
+                    condition_function=validators.oneOf((1, 2)),
+                    result_field="WORK_PART_STATUS",
+                    result_function=validators.oneOf(
+                        ["01", "02", "05", "07", "09", "15", "17", "18", "19", "99"]
```
While I do like condensing this line, all these formatting changes drive up the chances of mistakes uncaught in review.

andrew-jameson commented on tdrs-backend/tdpservice/parsers/test/test_parse.py:
```diff
@@ -888,3 +889,47 @@ def test_parse_tanf_section4_file(tanf_section4_file):
+    """Test parsing Tribal TANF Section 1 submission."""
+    parse.parse_datafile(tribal_section_1_file)
+
+    assert Tribal_TANF_T1.objects.all().count() == 121
+    assert Tribal_TANF_T2.objects.all().count() == 199
+    assert Tribal_TANF_T3.objects.all().count() == 213
```
Didn't we posit that hardcoding numbers like this leads to a maintenance tax for future changes? IIRC we didn't want to do so. @raftmsohani

raftmsohani replied:
we should parse a file that has only one T1, T2 and T3 and assert some of the parsed values. Parsing a large file doesn't help in this case

elipe17 replied:
I thought we only wanted that for the parser errors since there is not an easy way to determine the exact number of errors in each file? We can always figure out how many records of each type are in a file with a quick grep.

raftmsohani replied:
@elipe17 I am still not clear on why we need to parse such a large file. We should take minimal approach and create a test file that has only the relevant section 1 lines, it will be easier for us to look into the file and assert values

andrew-jameson commented on tdrs-backend/tdpservice/parsers/util.py:
```diff
@@ -103,10 +103,17 @@ def update_encrypted_fields(self, is_encrypted):
+def contains_tribe_code(line, tribe_code_val):
+    """Determine if the line has a valid tribal code."""
+    options = {"   ", None, "000"}
+    if tribe_code_val is not None:
```
I'm confused on checking for `None` but then it exists in `options`. Can you elaborate?

elipe17 replied:
Just me not thinking. The branch is unnecessary!

andrew-jameson commented on tdrs-backend/tdpservice/parsers/validators.py:
```diff
@@ -375,3 +375,25 @@ def validate_header_section_matches_submission(datafile, section, generate_error
+    if program_type == 'TAN' and (state_fips_code == '00' or state_fips_code is None or state_fips_code.isspace()):
+        is_valid = tribe_code is not None and not tribe_code.isspace() and tribe_code != '000'
+    else:
+        is_valid = tribe_code is None or tribe_code.isspace() or tribe_code == '000'
```
Is this redundant with `util.contains_tribe_code()`?

elipe17 replied:
Yes it was!

andrew-jameson commented on tdrs-backend/tdpservice/parsers/validators.py:
```diff
@@ -375,3 +375,25 @@ def validate_header_section_matches_submission(datafile, section, generate_error
+    is_valid = False
+
+    if program_type == 'TAN' and (state_fips_code == '00' or state_fips_code is None or state_fips_code.isspace()):
+        is_valid = tribe_code is not None and not tribe_code.isspace() and tribe_code != '000'
+    else:
+        is_valid = tribe_code is None or tribe_code.isspace() or tribe_code == '000'
```
These expressions seem very very close and almost the same check is being done against the `state_fips_code` in the conditional. I think refactoring here could lead to increased readability.

elipe17 replied:
I generalized a function to condense the code.

andrew-jameson reviewed (changes requested):
Have some minor changes requested and some questions.

raftmsohani commented on tdrs-backend/tdpservice/parsers/parse.py:
```diff
@@ -28,11 +28,29 @@ def parse_datafile(datafile):
+    is_tribal = util.contains_tribe_code(header_line, field_values["tribe_code"])
+
     logger.debug(f"Datafile has encrypted fields: {is_encrypted}.")
+    logger.debug(f"Datafile is Tribal: {is_tribal}.")
```
Nit: add information about the datafile in .debug, maybe the "DataFile.\_\_repr\_\_"

sugg: logger.debug(f"Datafile {DataFile.\_\_repr\_\_} is Tribal: {is_tribal}.")

raftmsohani commented on tdrs-backend/tdpservice/parsers/row_schema.py:
```diff
@@ -165,9 +165,10 @@ def run_postparsing_validators(self, instance, generate_error):
 
         return is_valid, errors
 
-    def get_field_by_name(self, name):
```
I have used this function in my PR

elipe17 replied:
I will re-add for compatibility.

raftmsohani committed:
removed excessive space in validator

raftmsohani committed:
linting

raftmsohani committed:
Merge branch 'develop' into 2599-error-report-corrections-impact-wpr-metrics

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into 1114-tribal-s1

elipe17 committed:
- Updated based on review feedback

elipe17 committed:
- removing debug assert

elipe17 committed:
- Fix lint

raftmsohani committed:
listing

raftmsohani committed:
Merge branch 'develop' into 2599-error-report-corrections-impact-wpr-metrics

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into 1114-tribal-s1

elipe17 committed:
- deconflicting migrations

elipe17 committed:
- Fixing test

raftmsohani committed:
added m6

raftmsohani committed:
lint

raftmsohani committed:
corrected new line break

raftmsohani committed:
refactored validator logic

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into 1114-tribal-s1

elipe17 committed:
- resolving conflict

elipe17 committed:
- Fix lint errors

elipe17 committed:
- Resolve conflict

George-Hudson commented on tdrs-backend/tdpservice/parsers/schema_defs/header.py:
```diff
@@ -63,63 +63,14 @@
+                    validators.isInStringRange(15, 42),
+                    validators.isInStringRange(44, 51),
+                    validators.isInStringRange(53, 56),
+                    validators.oneOf(["66", "72", "78"]),
```
ty. my eyes appreciate this.

George-Hudson commented on tdrs-backend/tdpservice/search_indexes/documents/tribal.py:
```diff
@@ -1 +1,186 @@
+            'OTHER_NON_SANCTION',
+            'WAIVER_EVAL_CONTROL_GRPS',
+            'FAMILY_EXEMPT_TIME_LIMITS',
+            'FAMILY_NEW_CHILD'
```
I don't know our standards on this, but here we don't have a trailing comma and on line 143, we do.

George-Hudson reviewed (approved):
lgtm

jtimpe commented on tdrs-backend/tdpservice/parsers/util.py:
```diff
@@ -178,7 +176,16 @@ def get_schema_options(program, section, query=None, model=None, model_name=None
+            'A': {
+                'section': DataFile.Section.TRIBAL_ACTIVE_CASE_DATA,
+                'models': {
+                    'T1': schema_defs.tribal_tanf.t_t1,
```
minor nitpick, but i feel the naming of the modules could be updated so that the import is a little cleaner, like `schema_defs.tribal.t1`

jtimpe commented on tdrs-backend/tdpservice/parsers/util.py:
```diff
@@ -103,11 +103,9 @@ def update_encrypted_fields(self, is_encrypted):
-    if encryption_field is not None:
-        return encryption_field.parse_value(line) == "E"
-    return False
+def is_string_field_valid(val, field_length):
```
due to circular imports, i had to move some functions like this (pure, validator-ish functions that don't import from our application code) from `util` to `validators`, namely `value_is_empty`.

i'd recommend either moving this for consistency, or perhaps even parameterizing the `empty_values` array in `value_is_empty` to include a `0`-fill case since the function is so similar

elipe17 replied:
Going to wait until your work (#2748) merges, that way I can update based off of this feedback.

jtimpe commented on tdrs-backend/tdpservice/parsers/test/test_parse.py:
```diff
@@ -984,3 +985,47 @@ def test_parse_ssp_section3_file(ssp_section3_file):
+
+    t_t1_objs = Tribal_TANF_T1.objects.all().order_by("CASH_AMOUNT")
+    t_t2_objs = Tribal_TANF_T2.objects.all().order_by("MONTHS_FED_TIME_LIMIT")
+    t_t3_objs = Tribal_TANF_T3.objects.all().order_by("EDUCATION_LEVEL")
```
i noticed that the case aggregates columns in the frontend don't populate for tribal submissions
![image](https://github.com/raft-tech/TANF-app/assets/111305129/367397a0-36cb-4416-a2b0-72dabf92fa4a)

i double checked ssp and tanf section 1 and both work. it's not immediately clear to me why tribal does not, it could be an unfinished implementation of the `get_prog_from_section` util

jtimpe reviewed (changes requested):
parsing and validation in general look good to me, but i think we should fix the case aggregates before merging if possible, and include a unit test case for tribal aggregates.

the other comments i could go either way on.

raftmsohani committed:
Merge branch 'develop' into 2599-error-report-corrections-impact-wpr-metrics

raftmsohani committed:
linting and correction on t1

raftmsohani committed:
Merge branch 'develop' into 2599-error-report-corrections-impact-wpr-metrics

raftmsohani committed:
friendly_name correction from comments

raftmsohani committed:
Merge branch '2599-error-report-corrections-impact-wpr-metrics' of https://github.com/raft-tech/TANF-app into 2599-error-report-corrections-impact-wpr-metrics

raftmsohani committed:
friendly_name correction

raftmsohani committed:
corrected failing test for m5

raftmsohani committed:
refactor the field_json creation DRY

elipe17 committed:
- Renamed tribal files
- Updated to allow case aggregates
- Updated datafile to have a more convenient date range
- Updated tests to capture case aggregates

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into 1114-tribal-s1

elipe17 committed:
- Reduced file size
- Updated test

elipe17 committed:
- Fix lint errors

raftmsohani reviewed (approved):
LGTM!

raftmsohani committed:
friendly_name corrections

raftmsohani committed:
Merge branch 'develop' into 2599-error-report-corrections-impact-wpr-metrics

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into 1114-tribal-s1

elipe17 committed:
Merge branch '2599-error-report-corrections-impact-wpr-metrics' of https://github.com/raft-tech/TANF-app into 1114-tribal-s1

elipe17 committed:
- Adding friendly names to tribal

elipe17 committed:
- Fixing typo

raftmsohani committed:
linting and cleaning errors

raftmsohani committed:
linting

raftmsohani committed:
correction on friendly_names

raftmsohani committed:
corrected friendly_name for test_util

raftmsohani committed:
Merge branch 'develop' into 2599-error-report-corrections-impact-wpr-metrics

raftmsohani committed:
correction child care - number of months

raftmsohani committed:
Merge branch '2599-error-report-corrections-impact-wpr-metrics' of https://github.com/raft-tech/TANF-app into 2599-error-report-corrections-impact-wpr-metrics

George-Hudson committed:
fixed a few more typos and some spacing. (#2767)

* fixed a few more typos and some spacing.

* fixed linting issues

* missed a spot.

---------

Co-authored-by: George Hudson <ghudson@teamraft.com>

raftmsohani committed:
Merge branch 'develop' into 2599-error-report-corrections-impact-wpr-metrics

andrew-jameson commented on scripts/zap-scanner.sh:
```diff
@@ -112,10 +112,6 @@ ZAP_CLI_OPTIONS="\
   -config globalexcludeurl.url_list.url\(14\).description='Site - FontAwesome.com' \
   -config globalexcludeurl.url_list.url\(14\).enabled=true \
 
-  -config globalexcludeurl.url_list.url\(15\).regex='^https:\/\/.*\.cloud.gov\/.*$' \
-  -config globalexcludeurl.url_list.url\(15\).description='Site - Cloud.gov' \
-  -config globalexcludeurl.url_list.url\(15\).enabled=true \
```
Is this a result of our ZAP PR being merged to develop? Would hate to undue these accidently.

andrew-jameson reviewed (approved):
Once the weirdness with ZAP/CORS is resolved, LGTM.

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into 1114-tribal-s1

elipe17 committed:
- knowledge center

elipe17 committed:
- Integrating Jans changes. Parametrizing values_is_empty

elipe17 committed:
Merge branch '2599-error-report-corrections-impact-wpr-metrics' of https://github.com/raft-tech/TANF-app into 1114-tribal-s1

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into 1114-tribal-s1

ADPennington commented:
@elipe17 can you check the deployment workflow on this ticket, please?

ADPennington commented on tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t2.py:
```diff
@@ -0,0 +1,740 @@
+                    endIndex=49,
+                    required=False,
+                    validators=[
+                        validators.isLargerThanOrEqualTo(0),
```
i made a mistake here. possible values are: ` 0-2," "`

ADPennington commented on tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t2.py:
```diff
@@ -0,0 +1,740 @@
+                    endIndex=55,
+                    required=False,
+                    validators=[
+                        validators.isInLimits(0, 9),
```
```suggestion
                        validators.isInLimits(0, 2),
```

ADPennington commented on tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t2.py:
```diff
@@ -0,0 +1,740 @@
+                    endIndex=65,
+                    required=False,
+                    validators=[
+                        validators.isInLimits(0, 9),
```
```suggestion
                        validators.isInLimits(0, 2),
```

ADPennington commented on tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t3.py:
```diff
@@ -0,0 +1,618 @@
+            validators=[validators.isAlphaNumeric()],
+        ),
+        Field(
+            item="67",
```
```suggestion
            item="66",
```

ADPennington replied:
starting here all the T3 item numbers are off by one. I made this correction in the workbook. sorry about that!

elipe17 committed:
- Update based on review feedback

elipe17 commented:
@ADPennington, Updated based on your feedback and got it deployed to QASP.

ADPennington commented on tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t1.py:
```diff
@@ -0,0 +1,668 @@
+        RowSchema(
+            model=Tribal_TANF_T1,
+            preparsing_validators=[
+                validators.hasLength(123),
```
@elipe17 after a lot of digging, it looks like [the record layout for tribal TANF section 1](https://www.acf.hhs.gov/sites/default/files/documents/ofa/tribal_layout_for_section1_0.pdf) is incorrect. all of the tribal files I tested with have T1, T2, and T3 record lengths of 122. 

Sorry for this mishap here. Can you make this change for each record type in this ticket and re-deploy to qasp env?

elipe17 replied:
@ADPennington fixed and redeployed to qasp

elipe17 committed:
- Updated length checks

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into 1114-tribal-s1

ADPennington commented on tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t2.py:
```diff
@@ -0,0 +1,740 @@
+                    result_field="MARITAL_STATUS",
+                    result_function=validators.isInLimits(1, 5),
+                ),
+                validators.if_then_validator(
```
noticing here that when this kind of error pops up that parent_minor_child data element is missing in the name columns.  
```
error_message:
if family affiliation :1 validator1 passed then PARENT_MINOR_CHILD None is not larger or equal to 1 and smaller or equal to 3.

Item Name:
,family affiliation

Internal Variable Name:
,FAMILY_AFFILIATION

```

ADPennington replied:
note to self and data team: this appears to be a common error among tribal section 1 files.

ADPennington replied:
@elipe17 did you see this one?

elipe17 replied:
@ADPennington I have a fix for this. But, I am going to put it into the Tribal S4 PR because it is going to introduce a new migration. To keep the dependencies clean and avoid any painful migration issues, putting it in S4 makes the most sense.

ADPennington replied:
ok will test this in #2762

ADPennington commented on tdrs-backend/tdpservice/search_indexes/models/tribal.py:
```diff
@@ -1 +1,190 @@
+    RecordType = models.CharField(max_length=156, null=True, blank=False)
+    RPT_MONTH_YEAR = models.IntegerField(null=True, blank=False)
+    CASE_NUMBER = models.CharField(max_length=11, null=True, blank=False)
+    FIPS_CODE = models.CharField(max_length=2, null=True, blank=False)
```
@elipe17 this should be tribe_code, and its stored in the header record, as shown below:
`HEADER20214A00142TAN1EU`

tribe code in the header above is `142`

ADPennington commented on tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t1.py:
```diff
@@ -0,0 +1,668 @@
+                    endIndex=8,
+                    required=True,
+                    validators=[
+                        validators.dateYearIsLargerThan(1998),
```
nit: i think we need a different validator for `YYYY` in this field. this wont affect qasp approval, but below is an edge case:

records with `RPT_MONTH_YEAR` == `202 00` raises the following error message: `reporting month year is required but a value was not provided` and it still gets stored in the db. OFA ultimately cant do anything with these records, so these shouldnt be parsed because we cant infer the reporting period.  :warning:

ADPennington replied:
#2793

ADPennington commented on tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t1.py:
```diff
@@ -0,0 +1,668 @@
+                    startIndex=8,
+                    endIndex=19,
+                    required=True,
+                    validators=[validators.isAlphaNumeric()],
```
also need a better validator for this field. [see comment here](https://github.com/raft-tech/TANF-app/issues/2681#issuecomment-1852328435)

ADPennington commented on tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t1.py:
```diff
@@ -0,0 +1,668 @@
+                    endIndex=22,
+                    required=False,
+                    validators=[
+                        validators.isNumber(),
```
cleanup needed. captured in #2749

ADPennington commented on tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t1.py:
```diff
@@ -0,0 +1,668 @@
+                Field(
+                    item="26AVII",
+                    name="OTHER_SANCTION",
+                    friendly_name="other, sanction",
```
```suggestion
                    friendly_name="other sanction",
```

ADPennington commented on tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t1.py:
```diff
@@ -0,0 +1,668 @@
+                Field(
+                    item="26CIV",
+                    name="OTHER_NON_SANCTION",
+                    friendly_name="other, non-sanction",
```
```suggestion
                    friendly_name="other non-sanction",
```

ADPennington commented on tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t2.py:
```diff
@@ -0,0 +1,740 @@
+                    endIndex=8,
+                    required=True,
+                    validators=[
+                        validators.dateYearIsLargerThan(1998),
```
#2793

ADPennington commented on tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t2.py:
```diff
@@ -0,0 +1,740 @@
+                    startIndex=8,
+                    endIndex=19,
+                    required=True,
+                    validators=[validators.isAlphaNumeric()],
```
> also need a better validator for this field. [see comment here](https://github.com/raft-tech/TANF-app/issues/2681#issuecomment-1852328435)

ADPennington commented on tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t2.py:
```diff
@@ -0,0 +1,740 @@
+                    required=True,
+                    validators=[
+                        validators.dateYearIsLargerThan(1900),
+                        validators.dateMonthIsValid(),
```
#2746

ADPennington commented on tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t2.py:
```diff
@@ -0,0 +1,740 @@
+                    startIndex=29,
+                    endIndex=38,
+                    required=True,
+                    validators=[validators.validateSSN()],
```
needs logic from #2771

ADPennington commented on tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t2.py:
```diff
@@ -0,0 +1,740 @@
+                Field(
+                    item="36E",
+                    name="RECEIVE_SSI",
+                    friendly_name="receives social security income",
```
```suggestion
                    friendly_name="receives SSI",
```

ADPennington commented on tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t2.py:
```diff
@@ -0,0 +1,740 @@
+                    type="number",
+                    startIndex=44,
+                    endIndex=45,
+                    required=True,
```
@elipe17 i missed this one. can also accept a space: " "

ADPennington commented on tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t2.py:
```diff
@@ -0,0 +1,740 @@
+                    type="string",
+                    startIndex=51,
+                    endIndex=53,
+                    required=True,
```
@elipe17 i missed this one too. can also allow spaces: "  "

ADPennington commented on tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t2.py:
```diff
@@ -0,0 +1,740 @@
+                    result_field="COOPERATION_CHILD_SUPPORT",
+                    result_function=validators.oneOf((1, 2, 9)),
+                ),
+                validators.validate__FAM_AFF__HOH__Fed_Time(),
```
error message language incorrect: #2792

ADPennington commented on tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t2.py:
```diff
@@ -0,0 +1,740 @@
+                    endIndex=53,
+                    required=True,
+                    validators=[
+                        validators.isInStringRange(1, 10),
```
```suggestion
                        validators.isInStringRange(0, 10),
```

ADPennington commented on tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t2.py:
```diff
@@ -0,0 +1,740 @@
+                    endIndex=94,
+                    required=False,
+                    validators=[
+                        validators.isInStringRange(0, 99),
```
@elipe17 only 2 possible values here: 00 and "  ".

ADPennington commented on tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t3.py:
```diff
@@ -0,0 +1,620 @@
+            endIndex=69,
+            required=True,
+            validators=[
+                validators.dateYearIsLargerThan(1950),
```
```suggestion
                validators.dateYearIsLargerThan(1900),
```

ADPennington replied:
#2746

ADPennington commented on tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t3.py:
```diff
@@ -0,0 +1,620 @@
+            startIndex=2,
+            endIndex=8,
+            required=True,
+            validators=[],
```
#2793

ADPennington commented on tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t3.py:
```diff
@@ -0,0 +1,620 @@
+            startIndex=8,
+            endIndex=19,
+            required=True,
+            validators=[validators.isAlphaNumeric()],
```
> also need a better validator for this field. [see comment here](https://github.com/raft-tech/TANF-app/issues/2681#issuecomment-1852328435)

ADPennington commented on tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t3.py:
```diff
@@ -0,0 +1,620 @@
+            endIndex=28,
+            required=True,
+            validators=[
+                validators.dateYearIsLargerThan(1950),
```
```suggestion
                validators.dateYearIsLargerThan(1900),
```

ADPennington replied:
#2746

ADPennington replied:
recommended 1900 just for consistency in the codebase across sections 1 and 2, the latter of which does not distinguish between children and adults.

ADPennington commented on tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t3.py:
```diff
@@ -0,0 +1,620 @@
+            startIndex=28,
+            endIndex=37,
+            required=True,
+            validators=[validators.validateSSN()],
```
needs logic from #2771

ADPennington commented on tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t3.py:
```diff
@@ -0,0 +1,620 @@
+            startIndex=69,
+            endIndex=78,
+            required=True,
+            validators=[validators.validateSSN()],
```
needs logic from #2771

ADPennington commented on tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t3.py:
```diff
@@ -0,0 +1,620 @@
+        Field(
+            item="71B",
+            name="RECEIVE_SSI",
+            friendly_name="receives social security income",
```
```suggestion
            friendly_name="receives SSI",
```

ADPennington commented on tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t3.py:
```diff
@@ -0,0 +1,620 @@
+            startIndex=8,
+            endIndex=19,
+            required=True,
+            validators=[validators.isAlphaNumeric()],
```
> also need a better validator for this field. [see comment here](https://github.com/raft-tech/TANF-app/issues/2681#issuecomment-1852328435)

ADPennington commented on tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t3.py:
```diff
@@ -0,0 +1,620 @@
+            type="number",
+            startIndex=43,
+            endIndex=44,
+            required=True,
```
```suggestion
            required=False,
```

ADPennington commented on tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t3.py:
```diff
@@ -0,0 +1,620 @@
+            type="number",
+            startIndex=84,
+            endIndex=85,
+            required=True,
```
```suggestion
            required=False,
```

ADPennington commented on tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t3.py:
```diff
@@ -0,0 +1,620 @@
+            startIndex=2,
+            endIndex=8,
+            required=True,
+            validators=[],
```
#2793

ADPennington commented on tdrs-backend/tdpservice/parsers/schema_defs/header.py:
```diff
@@ -68,11 +68,15 @@
-                                  "40", "41", "42", "44", "45", "46", "47", "48", "49", "50", "51", "53",
-                                  "54", "55", "56", "66", "72", "78"))
+                validators.or_validators(
+                    validators.isInStringRange(0, 2),
```
we also need to allow spaces for `state_fips` (for tribes) as well as spaces for `tribe_code` (for states and territories).

ADPennington reviewed (changes requested):
@elipe17 this is in pretty good shape.  some testing notes and questions below. You'll notice that most of the comments reference other clean-up tickets. Let me know if it makes sense to keep those fixes in other tix. The only exception might be `CASE_NUMBER`. 

---

#### Header/Trailer
- tribes will often space-fill or zero-fill the `state_fips`. space-filling raises an error message: `state_fips is required but a value was not provided.` 
- currently not storing `FIPS_CODE` from header (`state_fips`) :question:
- currently not storing `TRIBE_CODE` from header (`tribe_code`) :question:
- how will OFA staff who query the db know which STTs each record is associated with? seems like we need the abovementioned fields in every schema? possibly a new ticket? :question:

#### T1, T2, and T3

- verified that tribal section 1 files have T1,T2, and T3 records have length == 122. this was implemented here :heavy_check_mark:
- records with `RPT_MONTH_YEAR` needs better validator for YYYY. #2793
- `CASE_NUMBER` validator needs an update to handle edgy case numbers. [see comment here](https://github.com/raft-tech/TANF-app/issues/2681#issuecomment-1852328435) 
- `COUNTY_FIPS_CODE` validation correct but also related to a bug in FTANF executables that erroneously drops leading zeros. #2749
- `DATE_OF_BIRTH` needs logic from #2746
- `SSN` needs logic from #2771
- a few suggestions/comments related to friendly names and range of values I missed. 

----
----

for @reitermb --please double check that these are captured in the hackmd:

- cat 2 error messages look like the following (#2693): 
  - `2 does not match 1.` 
  - `3 is not in [1, 2].`
  - `3 is not larger or equal to 1 and smaller or equal to 2.`
  - `family exempt time limits is required but a value was not provided.`
  - `04 is not in range [0, 3]. or 04 is not in range [5, 9]. or 04 is not in range [11, 19]. or 04 does not match 99.`

- cat 3 error mesages look like the following (#2792): 
  - `if cash amount :502 validator1 passed then number of months 0 is not larger than 0.` 
  - `if child care amount :3 validator1 passed then children covered 0 is not larger than 0.`
  - `if sanction reduction amount :3 validator1 passed then work requirement sanction 0 is not in (1, 2).`
  - `The sum of ('amount of food stamp assistance', 'amount of subsidized child care', 'cash amount', 'child care amount', 'transportation amount', 'transition services amount', 'other amount') is not larger than 0.`
  - `if family affiliation :1 validator1 passed then work participation status 04 is not in range [1, 3]. or 04 is not in range [5, 9]. or 04 is not in range [11, 19]. or 04 does not match 99.`
  - `If family affiliation ==2 and citizenship status==1 or 2, then social security number - ssn != 000000000 -- 999999999.`

elipe17 committed:
- removing fips code

elipe17 committed:
Update tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t2.py

Co-authored-by: Alex P.  <63075587+ADPennington@users.noreply.github.com>

elipe17 committed:
Update tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t3.py

Co-authored-by: Alex P.  <63075587+ADPennington@users.noreply.github.com>

elipe17 committed:
Update tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t3.py

Co-authored-by: Alex P.  <63075587+ADPennington@users.noreply.github.com>

elipe17 committed:
Update tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t3.py

Co-authored-by: Alex P.  <63075587+ADPennington@users.noreply.github.com>

elipe17 committed:
- update based on review

elipe17 committed:
Merge branch '1114-tribal-s1' of https://github.com/raft-tech/TANF-app into 1114-tribal-s1

elipe17 committed:
- Fix migration

elipe17 commented:
@ADPennington I updated based on your review. I think all the validator changes should live in those tickets you called out. That way we can update all sections for programs in one go. I would also like to make the fips_code/tribe_code addition to the models/documents it's own ticket. Adding those to the models that are missing them will create a new migration. The migration generated will be a big one since it will have to impute a new column and value for the DB tables that don't have it. Those tables are pretty big which will cause a reasonably long running migration.

ADPennington commented:
> @ADPennington I updated based on your review. I think all the validator changes should live in those tickets you called out. That way we can update all sections for programs in one go. I would also like to make the fips_code/tribe_code addition to the models/documents it's own ticket. Adding those to the models that are missing them will create a new migration. The migration generated will be a big one since it will have to impute a new column and value for the DB tables that don't have it. Those tables are pretty big which will cause a reasonably long running migration.

sounds good thank you @elipe17. we may need to separate new tickets:

- for fips and tribe code storage in the record
- to address the case number validator. currently that ticket is focused on TANF Section 1 but the validator is used in SSP and Tribal TANF.

ADPennington reviewed (commented):
@elipe17 this is ready for approval as soon as [this comment](https://github.com/raft-tech/TANF-app/pull/2742/files#r1435224022) is addressed.

elipe17 committed:
- Fixed name clash on PARENT_MINOR_CHILD

elipe17 committed:
Revert "- Fixed name clash on PARENT_MINOR_CHILD"

This reverts commit 5850dc845e2f84d7858e591cb6c0f5c381e63470.

ADPennington reviewed (approved):
thank you @elipe17 🚀
